File: src/steward/operator/util/gitOperation.go
package util

import (
	"bytes"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"go.uber.org/zap"
)

type GitUser struct{
	Name string
	Email string
	Token string
}

var credentialHelperScript = "/tmp/credentialHelper.sh"

func GitSetup(gitUser GitUser) {
	createCredHelperScript(gitUser.Token)
	out, err := execGitCommand([]string{"config", "--global", "user.email", gitUser.Email}, ".")
	if err != nil {
		zap.L().Warn("git config setup error", zap.String("output", out), zap.Error(err))
	}
	out, err = execGitCommand([]string{"config", "--global", "user.name", gitUser.Name}, ".")
	if err != nil {
		zap.L().Warn("git config setup error", zap.String("output", out), zap.Error(err))
	}
}

func createCredHelperScript(userToken string) {
	txt := []byte("printf '%s\\n' " + userToken)
	err := ioutil.WriteFile(credentialHelperScript, txt, 0710)
	if err != nil {
		zap.L().Error("create credential helper error", zap.Error(err))
	}
	zap.L().Debug("create credential helper script complete")
}

func gitCloneRepository(repoPath string, gitURL string) error {
	if _, err := os.Stat(filepath.Join(repoPath, ".git")); !os.IsNotExist(err) {
		zap.L().Debug(fmt.Sprintf("Repo [%v] existed, skipped clone [%v]", repoPath, gitURL))
		return nil
	}

	zap.L().Debug(fmt.Sprintf("Cloning Data from [%v] [%v]...", gitURL, repoPath))
	os.MkdirAll(repoPath, 0755)

	_, err := execGitPasswordCommand([]string{"clone", gitURL, repoPath}, repoPath)

	if err != nil {
		zap.L().Error(fmt.Sprintf("Clone fail [%v]", err))
		return err
	}

	return nil
}

func gitFetch(repoPath string) error {
	zap.L().Info(fmt.Sprintf("Fetch Data [%v]...", repoPath))
	_, err := execGitPasswordCommand([]string{"fetch"}, repoPath)
	if err != nil {
		return err
	}

	zap.L().Info("Fetch Succeed")
	return nil
}

func gitPull(repoPath string) error {
	zap.L().Info(fmt.Sprintf("Pulling Data [%v]...", repoPath))
	_, err := execGitPasswordCommand([]string{"pull"}, repoPath)
	if err != nil {
		return err
	}

	zap.L().Info("Pull Succeed")
	return nil
}

func gitCommitAll(repoPath string, message string) error {
	lfsCheck(repoPath)
	_, err := execGitCommand([]string{"add", "--all"}, repoPath)
	if err != nil {
		return err
	}
	_, err = execGitCommand([]string{"commit", "-a", "-m", message, "--allow-empty"}, repoPath)
	if err != nil {
		return err
	}
	return nil
}

func gitTag(repoPath string, tagName string) error {
	_, err := execGitCommand([]string{"tag", tagName}, repoPath)
	return err
}

func gitPushArgs(repoPath string, args []string) error {
	zap.L().Info(fmt.Sprintf("Pushing to [%v] args [%v]...", repoPath, args))

	_, err := execGitPasswordCommand(
		append([]string{"push", "origin"}, args...),
		repoPath,
	)
	if err != nil {
		return err
	}

	zap.L().Info("Push Succeed")
	return nil
}

func gitPushRefs(repoPath string, commits []string) error {
	return gitPushArgs(repoPath, commits)
}

func gitPushAll(repoPath string) error {
	return gitPushArgs(repoPath, []string{"--all"})
}

func gitPushTags(repoPath string) error {
	return gitPushArgs(repoPath, []string{"--tags"})
}

func gitBranch(repoPath string, branchName string) error {
	_, err := execGitCommand([]string{"branch", branchName}, repoPath)
	if err != nil {
		return err
	}
	return nil
}

func gitCheckout(repoPath string, ref string, branch string) error {
	args := []string{"checkout"}

	if ref != "" {
		args = append(args, ref)
	}

	if branch != "" {
		args = append(args, "-b", branch)
	}

	_, err := execGitCommand(args, repoPath)
	if err != nil {
		return err
	}
	return nil
}

func lfsCheck(repoPath string) {
	filepath.Walk(repoPath, processFile(repoPath))
}

func processFile(repoPath string) filepath.WalkFunc {
	return func(path string, f os.FileInfo, err error) error {
		if err != nil {
			zap.L().Fatal("", zap.Error(err))
		}

		// ignore .git directory
		if f.IsDir() && f.Name() == ".git" {
			return filepath.SkipDir
		}

		switch mode := f.Mode(); {
		case mode.IsDir():
			// skip directory
			return nil
		case mode.IsRegular():
			if !isTextFile(path) {
				err := lfsTrackFile(f.Name(), repoPath)
				if err != nil {
					zap.L().Fatal("git lfs track file error", zap.Error(err))
				}
			}
		}
		return nil
	}
}

// To Rewrite
func isTextFile(path string) bool {
	out, err := execCommand("file", []string{path}, ".", []string{})
	if err != nil {
		zap.L().Fatal("run file command fail", zap.Error(err))
	}
	return strings.Contains(out, "text") || strings.Contains(out, "JSON")
}

func lfsTrackFile(filename string, repoPath string) error {
	zap.L().Info("git lfs track file", zap.String("file", filename))
	_, err := execGitCommand([]string{"lfs", "track", filename}, repoPath)
	if err != nil {
		return err
	}
	return nil
}

func execGitPasswordCommand(args []string, path string) (string, error) {
	return execCommand(
		"git",
		args,
		path,
		append(os.Environ(), "GIT_ASKPASS=" + credentialHelperScript),
	)
}

func execCommand(name string, args []string, path string, env []string) (string, error) {
	cmd := exec.Command(name, args...)
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if path != "" {
		cmd.Dir = path
	}
	if len(env) > 0 {
		cmd.Env = env
	}
	err := cmd.Run()
	zap.L().Debug("exec command",
		zap.String("command", name+" "+strings.Join(args, " ")),
		zap.String("output", stdout.String()))
	if err != nil {
		return stderr.String(), err
	}

	return stdout.String(), nil
}

func execGitCommand(args []string, cwd string) (string, error) {
	return execCommand("git", args, cwd, []string{})
}

func removeAllLocalBranch(cwd string) (string, error) {
	// git for-each-ref --format '%(refname:short)' refs/heads | xargs git branch -D
	cmd1 := exec.Command("git", "for-each-ref", "--format", "%(refname:short)", "refs/heads")
	cmd2 := exec.Command("xargs", "git", "branch", "-D")

	if cwd != "" {
		cmd1.Dir = cwd
		cmd2.Dir = cwd
	}

	read, write := io.Pipe() 
	cmd1.Stdout = write
	cmd2.Stdin = read

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd2.Stdout = &stdout
	cmd2.Stderr = &stderr

	cmd1.Start()
	cmd2.Start()
	cmd1.Wait()
	write.Close()
	cmd2.Wait()

	return stdout.String(), nil
}

func gitCheckUpdatedBranches(repoPath string) (map[string]struct{}, error) {
	if err := retryOperation(func() error {
		zap.L().Debug("fetch")
		return gitFetch(repoPath)
	}); err != nil {
		return nil, err
	}

	branchOutputToSet := func (output string, prefix string) map[string]string {
		branchSet := map[string]string{}
		for _, branch := range(strings.Split(output, "\n")) {
			tmp := strings.Split(branch, " ")
			if len(tmp) != 2 {
				continue
			}
			if tmp[0][len(prefix):] == "HEAD" {
				continue
			}
			branchSet[tmp[0][len(prefix):]] = tmp[1]
		}

		return branchSet
	}

	getBranchSet := func (pattern string) (map[string]string, error) {
		branchOutput, err := execGitCommand([]string{"for-each-ref", "--format", "%(refname) %(objectname)", pattern}, repoPath)
		if err != nil {
			return nil, err
		}

		return branchOutputToSet(branchOutput, pattern), nil
	}

	remoteBranchSet, err := getBranchSet("refs/remotes/origin/")
	if err != nil {
		return nil, err
	}

	localBranchSet, err := getBranchSet("refs/heads/")
	if err != nil {
		return nil, err
	}

	updatedBranches := map[string]struct{}{}
	for branch, remoteCommit := range(remoteBranchSet) {
		localCommit, branchExist := localBranchSet[branch]
		if !branchExist || localCommit != remoteCommit {
			updatedBranches[branch] = struct{}{}
		}
	}

	return updatedBranches, nil
}

File: src/logserver/gitoperation.py
import re
import logging
import os
import subprocess
import stat


base_dir = '/repos'
credential_helper_script = "/tmp/credentialHelper.sh"
logger = logging.getLogger(__name__)


class ModelRepo():
    def __init__(self, git_http_url):
        self.git_http_url = git_http_url


class GitUser:
    def __init__(self, name: str, email: str, token: str):
        self.name = name
        self.email = email
        self.token = token


def setup_git(repo_path: str, git_user: GitUser):
    logger.info('setup git user')
    create_cred_helper_script(git_user.token)
    subprocess.run(['git', 'config', 'user.email',
                    git_user.email], cwd=repo_path, check=True)
    subprocess.run(['git', 'config', 'user.name', git_user.name],
                   cwd=repo_path, check=True)


def create_cred_helper_script(user_token: str):
    logger.info('create credential helper script')
    txt = "printf '%s\\n' " + user_token
    f = open(credential_helper_script, 'w')
    f.write(txt)
    f.close()
    os.chmod(credential_helper_script, stat.S_IRUSR |
             stat.S_IWUSR | stat.S_IXUSR)


def convert_git_http_url_to_full_name(git_http_url: str) -> str:
    pattern = r"(?P<method>https?):\/\/(?:[\w_-]+@)(?P<provider>.*?(?P<port>\:\d+)?)(?:\/|:)(?P<handle>(?P<owner>.+?)\/(?P<repo>.+?))(?:\.git|\/)?$"
    result = re.match(pattern, git_http_url)
    return '{}/{}'.format(result['owner'], result['repo'])


def get_repo_path(git_http_url: str) -> str:
    return os.path.join(base_dir, convert_git_http_url_to_full_name(git_http_url))


def clone_repo(repo_path: str, git_http_url: str):
    logger.info(f'cloning repo {git_http_url}...')
    if os.path.isdir(os.path.join(repo_path, '.git')):
        logger.debug(f'repo {git_http_url} already existed at {repo_path}')
        return

    repo_path = get_repo_path(git_http_url)
    exec_git_password_command(['clone', git_http_url, repo_path])


def fetch_repo(repo_path: str):
    logger.debug(f'fetch repo {repo_path}')
    exec_command('git', ['fetch'], path=repo_path)


def checkout_file(repo_path: str, tree_ish: str, filename: str):
    logger.debug(f'checkout file {filename} from repo: {repo_path} with ref: {tree_ish}')
    exec_git_password_command(
        ['checkout', tree_ish, '--', filename], path=repo_path)


def exec_git_password_command(args, path=None):
    env = os.environ.copy()
    env['GIT_ASKPASS'] = credential_helper_script
    exec_command('git', args, path, env=env)


def exec_command(name: str, args, path=None, env=os.environ.copy()):
    logging.debug(f'exec commands: {[name] + args}')
    try:
        p = subprocess.run([name] + args, cwd=path, env=env,
                           stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    except subprocess.SubprocessError as err:
        logging.fatal(f'exec commands error: {err}')
    if p.stderr:
        logging.fatal(f'command execution stderr: {p.stderr}')
    if p.stdout:
        logging.debug(f'execution stdout: {p.stdout}')
